// @syntest/module/dist/lib/extension/Command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Command = void 0;
/*
 * Copyright 2020-2023 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Core.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const logging_1 = require("@syntest/logging");
class Command {
    constructor(moduleManager, tool, name, description, options, handler) {
        this.builder = (yargs) => {
            for (const option of this.options.keys()) {
                yargs = yargs.option(option, this.options.get(option));
            }
            return yargs.usage(`Usage: $0 ${this.tool} ${this.command} [options]`);
        };
        this.tool = tool;
        this.command = name;
        this.describe = description;
        this.options = options;
        this.handler = async (arguments_) => {
            await handler(arguments_);
            const LOGGER = (0, logging_1.getLogger)("Command");
            LOGGER.info("Cleaning up...");
            await moduleManager.cleanup();
        };
    }
}
exports.Command = Command;
