// @syntest/search-javascript/dist/lib/testcase/statements/Statement.js
"use strict";
/*
 * Copyright 2020-2023 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Javascript.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Statement = void 0;
/**
 * @author Dimitri Stallenberg
 */
class Statement {
    get variableIdentifier() {
        return this._variableIdentifier;
    }
    get typeIdentifier() {
        return this._typeIdentifier;
    }
    get name() {
        return this._name;
    }
    get ownType() {
        return this._ownType;
    }
    get uniqueId() {
        return this._uniqueId;
    }
    /**
     * Constructor
     * @param identifierDescription
     * @param ownType
     * @param uniqueId
     */
    constructor(variableIdentifier, typeIdentifier, name, ownType, uniqueId) {
        this._variableIdentifier = variableIdentifier;
        this._typeIdentifier = typeIdentifier;
        this._name = name;
        this._ownType = ownType;
        this._uniqueId = uniqueId;
    }
}
exports.Statement = Statement;
