// @syntest/module/dist/lib/util/Configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = exports.OptionGroups = void 0;
const Yargs = require("yargs");
var OptionGroups;
(function (OptionGroups) {
    OptionGroups["General"] = "General Options:";
    OptionGroups["Module"] = "Module Options:";
})(OptionGroups = exports.OptionGroups || (exports.OptionGroups = {}));
exports.Configuration = {
    configureUsage() {
        return (Yargs.usage(`Usage: syntest <tool> <command> [options]`)
            // TODO examples
            .epilog("visit https://syntest.org for more documentation"));
    },
    configureOptions(yargs) {
        return yargs
            .option("config", {
            alias: ["c"],
            default: ".syntest.json",
            description: "The syntest configuration file",
            group: OptionGroups.General,
            hidden: false,
            config: true,
            type: "string",
        })
            .option("preset", {
            alias: [],
            default: "none",
            description: "The preset you want to use",
            group: OptionGroups.General,
            hidden: false,
            type: "string",
        })
            .option("modules", {
            alias: ["m"],
            array: true,
            default: [],
            description: "List of dependencies or paths to modules to load",
            group: OptionGroups.Module,
            hidden: false,
            type: "string",
        });
    },
};
