// @syntest/logging/dist/lib/Configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = exports.OptionGroups = void 0;
var OptionGroups;
(function (OptionGroups) {
    OptionGroups["Logging"] = "Logging Options:";
})(OptionGroups = exports.OptionGroups || (exports.OptionGroups = {}));
exports.Configuration = {
    configureOptions(yargs) {
        return (yargs
            // logging options
            .options("log-directory", {
            alias: [],
            default: "logs",
            description: "The path where the logs should be saved",
            group: OptionGroups.Logging,
            hidden: false,
            normalize: true,
            type: "string",
        })
            .options("console-log-level", {
            alias: [],
            choices: [
                "silent",
                "debug",
                "error",
                "warn",
                "info",
                "verbose",
                "silly",
            ],
            default: "error",
            description: "Log level of the tool",
            group: OptionGroups.Logging,
            hidden: false,
            type: "string",
        })
            .options("file-log-level", {
            alias: [],
            default: ["info", "warn", "error"],
            description: "Which levels should be logged to file",
            group: OptionGroups.Logging,
            hidden: false,
            type: "array",
        }));
    },
};
