// @syntest/metric/dist/lib/util/diagnostics.js
"use strict";
/*
 * Copyright 2023-2023 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Core.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldNeverHappen = exports.seriesDistributionTypeNotRegistered = exports.seriesDistributionNotRegistered = exports.seriesUnitNotRegistered = exports.seriesNotRegistered = exports.distributionNotRegistered = exports.propertyNotRegistered = void 0;
/**
 * This file is meant to provide consistent error messages throughout the tool.
 */
const propertyNotRegistered = (property) => `Cannot record property! Metric '${property}' is not registered by any module!`;
exports.propertyNotRegistered = propertyNotRegistered;
const distributionNotRegistered = (distribution) => `Cannot record distribution! Metric '${distribution}' is not registered by any module!`;
exports.distributionNotRegistered = distributionNotRegistered;
const seriesNotRegistered = (series) => `Cannot record series! Metric '${series}' is not registered by any module!`;
exports.seriesNotRegistered = seriesNotRegistered;
const seriesUnitNotRegistered = (seriesName, seriesUnit) => `Cannot record series! Metric '${seriesName}.${seriesUnit}' is not registered by any module!`;
exports.seriesUnitNotRegistered = seriesUnitNotRegistered;
const seriesDistributionNotRegistered = (distribution) => `Cannot record series distribution! Metric '${distribution}' is not registered by any module!`;
exports.seriesDistributionNotRegistered = seriesDistributionNotRegistered;
const seriesDistributionTypeNotRegistered = (distribution, seriesUnit) => `Cannot record series distribution! Metric '${distribution}.${seriesUnit}' is not registered by any module!`;
exports.seriesDistributionTypeNotRegistered = seriesDistributionTypeNotRegistered;
const shouldNeverHappen = (bugLocation) => `This should never happen.\nThere is likely a bug in the ${bugLocation}.`;
exports.shouldNeverHappen = shouldNeverHappen;
