// @syntest/search/dist/lib/SearchSubject.js
"use strict";
/*
 * Copyright 2020-2021 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Core.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchSubject = void 0;
/**
 * Subject of the search process.
 *
 * @author Mitchell Olsthoorn
 */
class SearchSubject {
    /**
     * Constructor.
     *
     * @param target Target of the subject
     * @param rootContext RootContext of the subject
     * @protected
     */
    constructor(target, rootContext) {
        this._target = target;
        this._rootContext = rootContext;
        this._objectives = new Map();
        this._extractObjectives();
    }
    /**
     * Retrieve objectives.
     */
    getObjectives() {
        return [...this._objectives.keys()];
    }
    /**
     * Retrieve child objectives.
     *
     * @param objective The objective to get the child objectives of
     */
    getChildObjectives(objective) {
        return [...this._objectives.get(objective)];
    }
    get name() {
        return this._target.name;
    }
    get cfg() {
        return this._rootContext.getControlFlowProgram(this.path);
    }
    get path() {
        return this._target.path;
    }
}
exports.SearchSubject = SearchSubject;
