// @syntest/search-javascript/dist/lib/testcase/statements/action/ConstantObject.js
"use strict";
/*
 * Copyright 2020-2023 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Javascript.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstantObject = void 0;
const analysis_javascript_1 = require("@syntest/analysis-javascript");
const prng_1 = require("@syntest/prng");
const ActionStatement_1 = require("./ActionStatement");
/**
 * @author Dimitri Stallenberg
 */
class ConstantObject extends ActionStatement_1.ActionStatement {
    constructor(variableIdentifier, typeIdentifier, name, uniqueId, export_) {
        super(variableIdentifier, typeIdentifier, name, analysis_javascript_1.TypeEnum.OBJECT, uniqueId, [], export_);
    }
    mutate(sampler, depth) {
        // delta mutations are non existance here so we make a copy instead
        return prng_1.prng.nextBoolean(sampler.deltaMutationProbability)
            ? this.copy()
            : sampler.constantObjectGenerator.generate(depth, this.variableIdentifier, this.typeIdentifier, this.export.id, this.name, sampler.statementPool);
    }
    copy() {
        return new ConstantObject(this.variableIdentifier, this.typeIdentifier, this.name, this.uniqueId, this.export);
    }
    decode(context) {
        const import_ = context.getOrCreateImportName(this.export);
        const decoded = `const ${context.getOrCreateVariableName(this)} = ${import_}`;
        return [
            {
                decoded: decoded,
                reference: this,
            },
        ];
    }
}
exports.ConstantObject = ConstantObject;
