// @syntest/search-javascript/dist/lib/testcase/statements/action/ConstructorCall.js
"use strict";
/*
 * Copyright 2020-2023 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Javascript.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructorCall = void 0;
const analysis_javascript_1 = require("@syntest/analysis-javascript");
const prng_1 = require("@syntest/prng");
const ActionStatement_1 = require("./ActionStatement");
/**
 * ConstructorCall
 */
class ConstructorCall extends ActionStatement_1.ActionStatement {
    get classIdentifier() {
        return this._classIdentifier;
    }
    /**
     * Constructor
     * @param type the return identifierDescription of the constructor
     * @param uniqueId optional argument
     * @param args the arguments of the constructor
     */
    constructor(variableIdentifier, typeIdentifier, classIdentifier, name, uniqueId, arguments_, export_) {
        super(variableIdentifier, typeIdentifier, name, analysis_javascript_1.TypeEnum.FUNCTION, uniqueId, arguments_, export_);
        this._classIdentifier = classIdentifier;
    }
    mutate(sampler, depth) {
        if (prng_1.prng.nextBoolean(sampler.deltaMutationProbability)) {
            const arguments_ = this.args.map((a) => a.copy());
            if (arguments_.length > 0) {
                const index = prng_1.prng.nextInt(0, arguments_.length - 1);
                arguments_[index] = arguments_[index].mutate(sampler, depth + 1);
            }
            return new ConstructorCall(this.variableIdentifier, this.typeIdentifier, this._classIdentifier, this.name, prng_1.prng.uniqueId(), arguments_, this.export);
        }
        else {
            return sampler.constructorCallGenerator.generate(depth, this.variableIdentifier, this.typeIdentifier, this.export.id, this.name, sampler.statementPool);
        }
    }
    copy() {
        const deepCopyArguments = this.args.map((a) => a.copy());
        return new ConstructorCall(this.variableIdentifier, this.typeIdentifier, this._classIdentifier, this.name, this.uniqueId, deepCopyArguments, this.export);
    }
    decode(context) {
        const argumentsDecoding = this.args.flatMap((a) => a.decode(context));
        const arguments_ = this.args
            .map((a) => context.getOrCreateVariableName(a))
            .join(", ");
        const import_ = context.getOrCreateImportName(this.export);
        const decoded = `const ${context.getOrCreateVariableName(this)} = new ${import_}(${arguments_})`;
        return [
            ...argumentsDecoding,
            {
                decoded: decoded,
                reference: this,
            },
        ];
    }
}
exports.ConstructorCall = ConstructorCall;
