// @syntest/search-javascript/dist/lib/testcase/statements/action/Setter.js
"use strict";
/*
 * Copyright 2020-2023 Delft University of Technology and SynTest contributors
 *
 * This file is part of SynTest Framework - SynTest Javascript.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Setter = void 0;
const analysis_javascript_1 = require("@syntest/analysis-javascript");
const prng_1 = require("@syntest/prng");
const ClassActionStatement_1 = require("./ClassActionStatement");
/**
 * @author Dimitri Stallenberg
 */
class Setter extends ClassActionStatement_1.ClassActionStatement {
    /**
     * Constructor
     * @param identifierDescription the return type options of the function
     * @param type always void
     * @param uniqueId id of the gene
     * @param property the name of the property
     * @param arg the argument of the setter
     */
    constructor(variableIdentifier, typeIdentifier, name, uniqueId, argument, constructor_) {
        super(variableIdentifier, typeIdentifier, name, analysis_javascript_1.TypeEnum.FUNCTION, uniqueId, [argument], constructor_);
    }
    mutate(sampler, depth) {
        let argument = this.args.map((a) => a.copy())[0];
        let constructor_ = this.constructor_.copy();
        if (prng_1.prng.nextBoolean(0.5)) {
            argument = argument.mutate(sampler, depth + 1);
        }
        else {
            constructor_ = constructor_.mutate(sampler, depth + 1);
        }
        return new Setter(this.variableIdentifier, this.typeIdentifier, this.name, prng_1.prng.uniqueId(), argument, constructor_);
    }
    copy() {
        const deepCopyArgument = this.args.map((a) => a.copy())[0];
        return new Setter(this.variableIdentifier, this.typeIdentifier, this.name, this.uniqueId, deepCopyArgument, this.constructor_.copy());
    }
    decode(context) {
        const constructorDecoding = this.constructor_.decode(context);
        const argumentDecoding = this.args.flatMap((a) => a.decode(context));
        const argument = this.args
            .map((a) => context.getOrCreateVariableName(a))
            .join(", ");
        const decoded = `${context.getOrCreateVariableName(this.constructor_)}.${this.name} = ${argument}`;
        return [
            ...constructorDecoding,
            ...argumentDecoding,
            {
                decoded: decoded,
                reference: this,
            },
        ];
    }
}
exports.Setter = Setter;
