// @syntest/storage/dist/lib/util/Configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = exports.OptionGroups = void 0;
var OptionGroups;
(function (OptionGroups) {
    OptionGroups["General"] = "General Options:";
    OptionGroups["Storage"] = "Storage Options:";
})(OptionGroups = exports.OptionGroups || (exports.OptionGroups = {}));
exports.Configuration = {
    configureOptions(yargs) {
        return (yargs
            .option("fid", {
            alias: [],
            default: "",
            description: "The syntest flow identifier",
            group: OptionGroups.General,
            hidden: true,
            type: "string",
        })
            // storage
            .options("syntest-directory", {
            alias: [],
            default: "syntest",
            description: "The path where everything should be saved",
            group: OptionGroups.Storage,
            hidden: false,
            normalize: true,
            type: "string",
        })
            .options("temp-syntest-directory", {
            alias: [],
            default: ".syntest",
            description: "The path where all temporary files should be saved",
            group: OptionGroups.Storage,
            hidden: false,
            normalize: true,
            type: "string",
        }));
    },
};
